### DISCUSSION TITLE
빈 스코프에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by nayonsoso** [2024-woowacourse-study/level-interview#134] [Spring Core]
참고 : https://prolog.techcourse.co.kr/quizzes/70/essay-answers

**Comment by alstn113** [comment]
빈 스코프(Bean Scope)는 스프링 프레임워크에서 빈(Bean) 객체의 생명주기와 인스턴스화 범위를 정의하는 방법입니다. 
각각의 빈 스코프는 빈 객체가 생성되고, 관리되며, 소멸되는 방식을 결정합니다.
스프링에서는 싱글톤, 프로토타입, 웹 관련 빈 스코프 등이 있습니다.

### 주요 빈 스코프 종류

- Singleton
    - 스프링 컨테이너의 시작과 종료까지 유지되는 가장 넓은 범위의 스코프이다.
    - 모든 요청에 동일한 빈 인스턴스가 반환됩니다.
- Prototype
    - 요청할 때마다 새로운 빈 인스턴스가 생성됩니다.
    - 각각의 요청에 대해 별도의 빈 인스턴스가 반환됩니다.
    - 스프링 컨테이너는 프로토타입 빈의 생성과 의존관계 주입 그리고 초기화까지만 관여한다. 
    - 종료 메서드가 호출되지 않는다.
    - 싱글톤 빈과 프로토타입 빈을 함께 사용할 때 프로토타입 빈이 새로 생성되지 않는 문제가 발생할 수 있다. 
    이 문제를 해결하기 위해서는 `ObjectProvider`나 `Provider`를 사용해야 한다.
- Request
    - 웹 애플리케이션에서 사용되며, HTTP 요청 당 하나의 빈 인스턴스가 생성됩니다.
    - 각 HTTP 요청마다 별도의 빈 인스턴스가 반환됩니다.
- Session, Application, WebSocket도 Request랑 비슷한 개념이다.  참고로 `@RequestScope`, `@SessionScope`라는 어노테이션도 있다.

**Comment by 3Juhwan** [comment]
스프링 프레임워크에서 빈 스코프는 스프링 컨테이너가 관리하는 빈의 라이프사이클과 가시성을 정의합니다. 빈 스코프는 `@Scope`로 설정할 수 있습니다. 

**싱글톤** 

- 기본 스코프입니다. 
- 스프링 컨테이너의 시작과 종료까지 유지되는 가장 넓은 범위의 스코프입니다. 
- 스프링 컨테이너당 하나의 빈 인스턴스만 생성됩니다. 

**프로토타입**

- 스프링 컨테이너는 프로토타입 빈의 생성과 의존관계 주입, 초기화까지만 관여하고 더는 관리하지 않는 매우 짧은 범위의 스코프입니다. 
- 프로토타입 스코프를 스프링 컨테이너에 조회하면 스프링 컨테이너는 항상 새로운 인스턴스를 생성해서 반환합니다. 
- 빈 관리의 책임이 클라이언트에게 있음으로 `@PreDestroy`가 실행되지 않습니다. 

**웹 관련 스코프**

- request: 웹 요청이 들어오고 나갈 때까지 유지되는 스코프입니다. 
- session: 웹 세션이 생성되고 종료될 때까지 유지되는 스코프입니다. 
- application: 웹의 서블릿 컨텍스트와 같은 범위로 유지되는 스코프입니다. 
- websocket: 웹 소켓과 동일한 생명주기를 가지는 스코프입니다.
